src/cast_core/src/SamplingLoader.cpp

```diff
@@ -48,67 +48,84 @@ using bsoncxx::builder::basic::make_document;
 /** @private */
 struct SamplingLoader::PhaseConfig {
     PhaseConfig(PhaseContext& context, mongocxx::pool::entry& client, ActorId id)
-        : database{(*client)[context["Database"].to<std::string>()]},
-          collection{database[context["Collection"].to<std::string>()]},
-          insertBatchSize{context["InsertBatchSize"].to<IntegerSpec>()},
-          numBatches{context["Batches"].to<IntegerSpec>()},
-          // A sample size is optional. If unspecified, we'll make a sample size such that each
-          // sample document is inserted one time.
-          sampleSize{
-              context["SampleSize"].maybe<IntegerSpec>().value_or(numBatches * insertBatchSize)},
-          pipelineSuffixGenerator{context["Pipeline"]
-                                      .maybe<PipelineGenerator>(context, id)
-                                      .value_or(PipelineGenerator{})} {}
-
-    mongocxx::database database;
-    mongocxx::collection collection;
+        : insertBatchSize{context["InsertBatchSize"].to<IntegerSpec>()},
+          numBatches{context["Batches"].to<IntegerSpec>()} {}
+
     // See src/workloads/docs/SamplingLoader.yml for a description of the arguments and some
     // examples.
     int64_t insertBatchSize;
     int64_t numBatches;
-    int64_t sampleSize;
-    PipelineGenerator pipelineSuffixGenerator;
 };
 
+std::vector<bsoncxx::document::value> DeferredSample::getSample() {
+    std::lock_guard<std::mutex> lock(_mutex);
+    if (_sampleDocs.empty()) {
+        _sampleDocs = gatherSample(lock);
+    }
+    return _sampleDocs;
+}
+
+std::vector<bsoncxx::document::value> DeferredSample::gatherSample(
+    const std::lock_guard<std::mutex>& lock) {
```

Comment:
Out of curiosity, why is the lock guard passed in? It doesn't seem used here, so perhaps this is a C++-ism I'm not familiar with?

Reply:
Oh yeah this was a quick and dirty carry-over of the "WithLock" concept in our server codebase: https://github.com/mongodb/mongo/blob/42667e3c762bd83d700696ec4635e41fb8471c0f/src/mongo/util/concurrency/with_lock.h#L40

I didn't carry over that type, but the intention was the same: make it impossible to call the function without holding the lock.

That being said - after typing this out. This function does not particularly _need_ to be called in the lock. The dangerous part is the assignment to _sampleDocs in the calling code.

Reply:
Coming back again - it does still feel worthwhile to protect this, since we use the pipeline suffix generator which is not thread safe. I didn't change anything.